go-controller/pkg/crd/clusteruserdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,107 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	udnv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterUserDefinedNetwork describe network request for a shared network across namespaces.
+//
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=clusteruserdefinednetworks,scope=Cluster
+// +kubebuilder:singular=clusteruserdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type ClusterUserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ClusterUserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status ClusterUserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+	// NamespaceSelector Label selector for which namespace network should be available for.
+	// +kubebuilder:validation:Required
+	// +required
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector"`
+
+	// Network is the user-defined-network spec
+	// +kubebuilder:validation:Required
+	// +required
+	Network NetworkSpec `json:"network"`
+}
+
+// NetworkSpec defines the desired state of UserDefinedNetworkSpec.
+// +union
+type NetworkSpec struct {
+	// Topology describes network configuration.
+	//
+	// Allowed values are "Layer3", "Layer2".
+	// Layer3 topology creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets.
+	// Layer2 topology creates one logical switch shared by all nodes.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	// +unionDiscriminator
+	Topology udnv1.NetworkTopology `json:"topology"`
+
+	// Layer3 is the Layer3 topology configuration.
+	// +optional
+	Layer3 *udnv1.Layer3Config `json:"layer3,omitempty"`
+
+	// Layer2 is the Layer2 topology configuration.
+	// +optional
+	Layer2 *udnv1.Layer2Config `json:"layer2,omitempty"`
+}
+
+// NamespaceNetworkStatus describe a user-defined-network status in a namespace who is part of a cluster
+// user-defined-network shared network.
+type NamespaceNetworkStatus struct {
+	// Name is the target namespace name
+	Name string `json:"name"`
+
+	// ActiveSinceTime indicate since when the network is available.
+	ActiveSinceTime metav1.Time `json:"activeSinceTime,omitempty"`
```

Comment:
Is there a reason to have a special struct to report per-namespace status instead of using per-namespace condition in the status.conditions? I see you mentioned some discussion https://github.com/openshift/enhancements/pull/1638/files#r1712960977 maybe you can point me to that?

Reply:
Currently, only the namespace name is required.
Following feedback on the suggested API and discussion offline it was concluded to avoid using bare strings in order to enable the API evolve and reflect more details.